samples/tutorials/mnist/04_Reusable_and_Pre-build_Components_as_Pipeline.ipynb

```diff
@@ -0,0 +1,799 @@
+      "  warnings.warn(_CLOUD_SDK_CREDENTIALS_WARNING)\n",
+      "/Users/luoshixin/LocalSim/virtualPython35/lib/python3.5/site-packages/google/auth/_default.py:66: UserWarning: Your application has authenticated using end user credentials from Google Cloud SDK. We recommend that most server applications use service accounts instead. If your application continues to use end user credentials from Cloud SDK, you might receive a \"quota exceeded\" or \"API not enabled\" error. For more information about service accounts, see https://cloud.google.com/docs/authentication/\n",
+      "  warnings.warn(_CLOUD_SDK_CREDENTIALS_WARNING)\n"
+     ]
+    },
+    {
+     "data": {
+      "text/plain": [
+       "'gcr.io/kubeflow-pipeline-fantasy/mnist_training_kf_pipeline@sha256:6ded6029bdd5c9dd0c95aee9e8fc81d0259f81088da9f1d68daf1bf517a329b7'"
+      ]
+     },
+     "execution_count": 7,
+     "metadata": {},
+     "output_type": "execute_result"
+    }
+   ],
+   "source": [
+    "IMAGE_NAME=\"mnist_training_kf_pipeline\"\n",
+    "TAG=\"latest\" # \"v_$(date +%Y%m%d_%H%M%S)\"\n",
+    "\n",
+    "GCR_IMAGE=\"gcr.io/{PROJECT_ID}/{IMAGE_NAME}:{TAG}\".format(\n",
+    "    PROJECT_ID=PROJECT_ID,\n",
+    "    IMAGE_NAME=IMAGE_NAME,\n",
+    "    TAG=TAG\n",
+    ")\n",
+    "\n",
+    "builder = kfp.containers._container_builder.ContainerBuilder(\n",
+    "    gcs_staging=GCS_BUCKET + \"/kfp_container_build_staging\")\n",
+    "\n",
+    "image_name = kfp.containers.build_image_from_working_dir(\n",
+    "    image_name=GCR_IMAGE,\n",
+    "    working_dir='./tmp/reuse_components_pipeline/mnist_training/',\n",
+    "    builder=builder\n",
+    ")\n",
+    "\n",
+    "image_name"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "#### If you want to use docker to build the image\n",
+    "Run the following in a cell\n",
+    "```bash\n",
+    "%%bash -s \"{PROJECT_ID}\"\n",
+    "\n",
+    "IMAGE_NAME=\"mnist_training_kf_pipeline\"\n",
+    "TAG=\"latest\" # \"v_$(date +%Y%m%d_%H%M%S)\"\n",
+    "\n",
+    "# Create script to build docker image and push it.\n",
+    "cat > ./tmp/components/mnist_training/build_image.sh <<HERE\n",
+    "PROJECT_ID=\"${1}\"\n",
+    "IMAGE_NAME=\"${IMAGE_NAME}\"\n",
+    "TAG=\"${TAG}\"\n",
+    "GCR_IMAGE=\"gcr.io/\\${PROJECT_ID}/\\${IMAGE_NAME}:\\${TAG}\"\n",
+    "docker build -t \\${IMAGE_NAME} .\n",
+    "docker tag \\${IMAGE_NAME} \\${GCR_IMAGE}\n",
+    "docker push \\${GCR_IMAGE}\n",
+    "docker image rm \\${IMAGE_NAME}\n",
+    "docker image rm \\${GCR_IMAGE}\n",
+    "HERE\n",
+    "\n",
+    "cd tmp/components/mnist_training\n",
+    "bash build_image.sh\n",
+    "```\n",
+    "\n",
+    "**Remember to set the image_name after the image is built**\n",
+    "```python\n",
+    "image_name = <the image uri>\n",
+    "```"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "## Writing your component definition file\n",
```

Comment:
Please apply the same feedback as notebook 03

Reply:
This has been changed to align with previous notebooks.